src/analyses/variable-sensitivity/constant_array_abstract_object.cpp

```diff
@@ -0,0 +1,580 @@
+
+  if(!stack.empty())
+  {
+    mp_integer index_value;
+    if(eval_index(index_expr, environment, ns, index_value))
+    {
+      // We were able to evaluate the index to a value, which we
+      // assume is in bounds...
+      auto const old_value = map.find(index_value);
+
+      if(!old_value.has_value())
+      {
+        result->map.insert(
+          index_value,
+          environment.write(
+            get_top_entry(environment, ns), value, stack, ns, merging_write));
+      }
+      else
+      {
+        result->map.replace(
+          index_value,
+          environment.write(
+            old_value.value(), value, stack, ns, merging_write));
+      }
+
+      result->clear_top();
+      DATA_INVARIANT(result->verify(), "Structural invariants maintained");
+      return result;
+    }
+    else
+    {
+      // We were not able to evaluate the index to a value
+      shared_array_mapt::viewt view;
+      map.get_view(view);
+
+      for(const auto &starting_value : view)
+      {
+        // Merging write since we don't know which index we are writing to
+        result->map.replace(
+          starting_value.first,
+          environment.write(starting_value.second, value, stack, ns, true));
+
+        result->clear_top();
+      }
+
+      DATA_INVARIANT(result->verify(), "Structural invariants maintained");
+      return result;
+    }
+  }
+  else
+  {
+    mp_integer index_value;
+
+    if(eval_index(index_expr, environment, ns, index_value))
+    {
+      // We were able to evalute the index expression to a constant
+      if(merging_write)
+      {
+        if(is_top())
+        {
+          DATA_INVARIANT(result->verify(), "Structural invariants maintained");
+          return result;
+        }
+
+        INVARIANT(!result->map.empty(), "If not top, map cannot be empty");
+
+        auto const old_value = result->map.find(index_value);
+
+        if(!old_value.has_value()) // Array element is top
+        {
+          DATA_INVARIANT(result->verify(), "Structural invariants maintained");
+          return result;
+        }
+
+        bool dummy;
+
+        result->map.replace(
+          index_value,
+          abstract_objectt::merge(old_value.value(), value, dummy));
+
+        DATA_INVARIANT(result->verify(), "Structural invariants maintained");
+        return result;
+      }
+      else
+      {
+        auto const old_value = result->map.find(index_value);
+        if(old_value.has_value())
+        {
+          if(value != abstract_object_pointert{old_value.value()})
+          {
+            result->map.replace(index_value, value);
+          }
+        }
+        else
+        {
+          result->map.insert(index_value, value);
+        }
+        result->clear_top();
+        DATA_INVARIANT(result->verify(), "Structural invariants maintained");
+        return result;
+      }
+    }
+
+    // try to write to all
+    // TODO(tkiley): Merge with each entry
```

Comment:
It's not that the current code is wrong, more that it could be improved.  If your index is not resolvable to a constant then there are a number of possible approaches.  Again, it is on my "to improve" list.